app/questionnaire/questionnaire_schema.py

```diff
@@ -0,0 +1,160 @@
+            return navigation.get('sections')
+
+        return None
+
+    def get_blocks(self):
+        for group in self.get_groups():
+            for block in group['blocks']:
+                yield block
+
+    def get_groups(self):
+        for group in self.json.get('groups', []):
+            yield group
+
+    def get_group(self, group_id):
+        return self._groups.get(group_id)
+
+    def get_block(self, block_id):
+        return self._blocks.get(block_id)
+
+    def get_answers_by_id_for_block(self, block_id):
+        answers = {}
+        block = self.get_block(block_id)
+        if block:
+            for question in block.get('questions', []):
+                for answer in question.get('answers', []):
+                    answers[answer['id']] = answer
+
+        return answers
+
+    def get_answer_ids_for_block(self, block_id):
+        return list(self.get_answers_by_id_for_block(block_id).keys())
+
+    def get_answer_schema_for_answer_id(self, block_id, answer_id):
+        answer_schema_list = self.get_answers_by_id_for_block(block_id)
+        if answer_id in answer_schema_list:
+            return answer_schema_list[answer_id]
+
+    def get_answers_for_block(self, block_id):
+        return list(self.get_answers_by_id_for_block(block_id).values())
+
+    def get_answers_that_repeat_in_block(self, block_id):
+        block = self.get_block(block_id)
+
+        for question in block.get('questions', []):
+            if question['type'] == 'RepeatingAnswer':
+                for answer in question['answers']:
+                    yield answer
+
+    def get_groups_that_repeat_with_answer_id(self, answer_id):
+        for group in self.get_groups():
+            repeating_rule = self.get_repeat_rule(group)
+            if repeating_rule and repeating_rule['answer_id'] == answer_id:
+                yield group
+
+    def get_summary_and_confirmation_blocks(self):
+        blocks = []
+
+        for block in self.get_blocks():
+            if block['type'] in ('Summary', 'Confirmation'):
+                blocks.append(block['id'])
+
+        return blocks
+
+    def get_parent_options_for_block(self, block_id):
+        options_with_children = {}
+
+        for answer_json in self.get_answers_for_block(block_id):
+            if answer_json['type'] in ['Checkbox', 'Radio']:
+                answer_options_with_children = {
+                    answer_json['id']: {
+                        'index': index,
+                        'child_answer_id': o['child_answer_id'],
+                    }
+                    for index, o in enumerate(answer_json['options']) if 'child_answer_id' in o}
+
+                options_with_children.update(answer_options_with_children)
+        return options_with_children
+
+    def group_has_questions(self, group_id):
+        for block in self.get_group(group_id)['blocks']:
+            if block['type'] == 'Questionnaire':
+                return True
+
+        return False
```

Comment:
Minor, but I think the methods should be grouped into the levels in which they query e.g. group level methods, block level methods, answer level methods